logstash-core/src/main/java/org/logstash/config/ir/compiler/EventCondition.java

```diff
@@ -0,0 +1,709 @@
+            }
+            return condition;
+        }
+
+        private static EventCondition in(final EventValueExpression left, final List<?> right) {
+            return new EventCondition.Compiler.FieldInConstantList(
+                PathCache.cache(left.getFieldName()), right
+            );
+        }
+
+        /**
+         * Compiles a constant (due to both of its sides being constant {@link ValueExpression})
+         * conditional into either {@link EventCondition.Compiler#TRUE} or
+         * {@link EventCondition.Compiler#FALSE}.
+         * @param left Constant left side {@link ValueExpression}
+         * @param right Constant right side {@link ValueExpression}
+         * @return Either {@link EventCondition.Compiler#TRUE} or
+         * {@link EventCondition.Compiler#FALSE}
+         */
+        private static EventCondition in(final ValueExpression left, final ValueExpression right) {
+            final Object found = right.get();
+            final Object other = left.get();
+            if (found instanceof ConvertedList && other instanceof RubyString) {
+                return ((ConvertedList) found).stream().anyMatch(item -> item.toString()
+                    .equals(other.toString())) ? TRUE : FALSE;
+            } else if (found instanceof RubyString && other instanceof RubyString) {
+                return found.toString().contains(other.toString()) ? TRUE : FALSE;
+            } else if (found instanceof RubyString && other instanceof ConvertedList) {
+                return ((ConvertedList) other).stream()
+                    .anyMatch(item -> item.toString().equals(found.toString())) ? TRUE : FALSE;
+            } else {
+                return found != null && other != null && found.equals(other) ? TRUE : FALSE;
+            }
+        }
+
+        private static boolean listValueRight(final In in) {
+            return ((ValueExpression) in.getRight()).get() instanceof List;
+        }
+
+        private static boolean scalarValueRight(final In in) {
+            return isScalar((ValueExpression) in.getRight());
+        }
+
+        private static boolean scalarValueLeft(final In in) {
+            return isScalar((ValueExpression) in.getLeft());
+        }
+
+        private static boolean isScalar(final ValueExpression expression) {
+            final Object value = expression.get();
+            return value instanceof String || value instanceof Number;
+        }
+
+        private static EventCondition in(final EventValueExpression left,
+            final EventValueExpression right) {
+            return new EventCondition.Compiler.FieldInField(
+                PathCache.cache(left.getFieldName()), PathCache.cache(right.getFieldName())
+            );
+        }
+
+        private static EventCondition eq(final EventValueExpression evale,
+            final ValueExpression vale) {
+            return new EventCondition.Compiler.FieldEquals(
+                evale.getFieldName(), vale.get().toString()
```

Comment:
@original-brownbear can we add a test for this too? It's worrying that we don't have one for different types of equality.